### PR TITLE
Provide iam policy for autoscaler addon

### PR DESCRIPTION
**Pull request by ese**
close: #1573
Add autoscaler [required permissions](https://github.com/kubernetes/contrib/tree/master/cluster-autoscaler/cloudprovider/aws) in the master policy and schedule the addon in masters

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1620)

### CODE REVIEW

**Comment by chrislovecnm** [comment]
CI failed. You probably have some terraform to update ;)

**Comment by chrislovecnm** [comment]
@k8s-bot ok to test

File: pkg/model/iam/iam_builder.go [hunk 1]
```diff
@@ -140,6 +140,17 @@ func (b *IAMPolicyBuilder) BuildAWSIAMPolicy() (*IAMPolicy, error) {
...
+                        Effect:   IAMStatementEffectAllow,
+                        Action:   []string{
+                                  "autoscaling:DescribeAutoScalingGroups",
+                                  "autoscaling:DescribeAutoScalingInstances",
+                                  "autoscaling:SetDesiredCapacity",
+                                  "autoscaling:TerminateInstanceInAutoScalingGroup",
+			},
+                        Resource: []string{"*"},
+                })
+
```
Comment (chrislovecnm) [inline pkg/model/iam/iam_builder.go]: How have you tested this?

Comment (ese) [inline pkg/model/iam/iam_builder.go]: I tested it manually

**Comment by ese** [comment]
make ci happy
close: #1621

**Comment by ese** [comment]
Maybe make sense run the addon in the kube-system namespace. What do you think?

**Comment by ese** [comment]
Rebased master due conflicts

**Comment by yissachar** [comment]
I am wondering if we really want to add this permission in kops. 

Not every cluster will run the cluster autoscaler, but with this PR they will all end up with those IAM permissions. I would prefer to keep the base set of IAM permissions as minimal as possible, and have users extend their particular cluster permissions as needed via our support for [adding additional IAM policies](https://github.com/kubernetes/kops/blob/master/docs/iam_roles.md#adding-additional-policies).

@justinsb @chrislovecnm Thoughts?

**Comment by kvz** [comment]
Can't speak for the k8s/kops developer standpoint, but as a user wanting to leverage scaling of nodes (and I can't think of many use cases that wouldn't benefit from having that), I would love for kops to take care of this - I kind of feel that's what kops is for: taking away as much friction as possible. As for the security perspective, kops can already create and destroy nodes with the credentials you provide it, so I don't think utilizing an autoscaler to do so will make things much worse than it already is in that regard(?)

**Comment by chrislovecnm** [comment]
@yissachar what is your concern about having this included?  We have the addon, and there is a bug in having the addon.  It does not work ... period.  

1. security concerns?
2. operational concerns?
3. concerns that it is not a base feature?
4. what is a better way to do this??

Service Catalog may be the answer.  It can talk to the aws api and make tweaks.  Before a deployment is installed.  Dependency management for the WIN

**Comment by yissachar** [comment]
@kvz Just to clarify, this permission goes on the master instances. This is separate from the credentials that are passed to kops.

@chrislovecnm I don't think there is a huge security concern here, it just seems to go against the current security approach that we seem to have in kops, which is that you only get the permissions that you need. For instance, if you're not using the encrypted volumes feature then you don't get the kms permissions. But for this feature, we would be giving all clusters the autoscaling permissions, regardless of if they need it or not.

The ideal would be if add-ons could be more managed by kops, and that way we could add the autoscaling permission as needed (I think this might be what you are referring to with Service Catalog, but I am not very familiar with this).

Either way, I don't have a super strong preference on this. If we decide that we want to prioritize the user convenience of setting up the autoscaler, I'm fine with that. On the other hand, I also don't think it's hugely inconvenient to tell the user to add those permissions in the addons doc.

**Comment by ese** [comment]
I think this is good solution for this iteration. I put the permissions in master thinking it is better in terms of security than having this policy in all nodes.

In the long term I think it would be better use kube2iam to isolate roles to the specific pods that need it but I think that requiere more deep develop and research to fit well all pieces, at least I dont know much of kube2iam yet. 
This will allow a more complex ecosystem of addons and provide flexibility to the users in their deployments

**Comment by justinsb** [comment]
I think that you're both right @yissachar  and @ese.  We should be using kube2iam, but a lot of people want this.  As the permissions are only on the master I don't think it's _too_ bad, so I'd like to allow this permission but remove it when we have kube2iam functionality.  That will be a breaking change, but removing the permission at that time will be pretty disruptive if we do it right anyway (we'll likely want to give fewer permissions to kube-apiserver, for example).

It sounds like you don't particularly object @yissachar, other than that we shouldn't continue to pile on more and more permissions - and I agree whole-heartedly with that.  We could have this be controlled by an option in the manifest, except that then we'd be coding in knowledge of the autoscaler into kops, and we'd likely want to remove it once we had kube2iam.

As such, I'm going to merge, and hopefully get some feedback on it in 1.5.0.

**Pull request merged**